### PR TITLE
v2.1.0

### PR DESCRIPTION
**Pull request by donavanbecker**
Just publish message to topic `homebridge/motion`

```{
  "platform": "Camera-ffmpeg",
  "mqtt": "127.0.0.1",
  "topics": "homebridge/motion",
  "cameras": [
    {
      "name": "Camera name",
      "motion": true,
      "videoConfig": {
        "source": "-re -i http://10.0.0.1",
        "stillImageSource": "-i http://10.0.0.1",
        "maxStreams": 5,
        "maxWidth": 1280,
        "maxHeight": 720,
        "maxFPS": 15,
        "maxBitrate": 1000
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by donavanbecker** [comment]
## v2.X.X (2020-XX-XX)

  ### Changes

  * Add MQTT support for Motion Detect (#572), thanks to [fennec622](https://github.com/fennec622).

**Comment by donavanbecker** [comment]
@fennec622 I think we will have to get this mqtt setup so that it can be set for each camera? if we don't, then it will only work for one camera. we can add a mqtt setting option for each camera. can you make those changes?

**Comment by fennec622** [comment]
No
You can one setup for all cameras 


Topic must like /homebridge/Motion/# dont forget #

And after 

If you have 
Camera1 camera2 ....

Just post /homebridge/Motion/camera1 for push camera1 etc etc

**Comment by donavanbecker** [comment]
> No
> You can one setup for all cameras
> 
> Topic must like /homebridge/Motion/# dont forget #
> 
> And after
> 
> If you have
> Camera1 camera2 ....
> 
> Just post /homebridge/Motion/camera1 for push camera1 etc etc

okay so if they put `#` it will work for all cameras?

**Comment by fennec622** [comment]
Yes 
with # you look all topics after homebridge/motion

when mqtt see homebridge/motion/camera1 motion detect on camera1
when mqtt see homebridge/motion/camera2 motion detect on camera2

So for that homebridge/motion/camera1 camera1 must exactly the same that accessory.displayName

**Comment by donavanbecker** [comment]
> Yes
> with # you look all topics after homebridge/motion
> 
> when mqtt see homebridge/motion/camera1 motion detect on camera1
> when mqtt see homebridge/motion/camera2 motion detect on camera2
> 
> So for that homebridge/motion/camera1 camera1 must exactly the same that accessory.displayName

Awesome! Would you test out the latest beta and let me know if it is working?

**Comment by fennec622** [comment]
Yes it's ok

**Comment by fennec622** [comment]
But i have a best way

Just publish message to topic homebridge/motion 
message is the name of camera with no space replace with _

`{
  "platform": "Camera-ffmpeg",
  "mqtt": "127.0.0.1",
  "topics": "homebridge/motion",
  "cameras": [
    {
      "name": "Camera name",
      "motion": true,
      "videoConfig": {
        "source": "-re -i http://10.0.0.1",
        "stillImageSource": "-i http://10.0.0.1",
        "maxStreams": 5,
        "maxWidth": 1280,
        "maxHeight": 720,
        "maxFPS": 15,
        "maxBitrate": 1000
      }
    }
  ]
}`

And modify index.ts like this

`var client = mqtt.connect("mqtt://"+servermqtt+":"+port);
      client.on('connect', () => {
        this.log("MQTT CONNECTED!")});
      client.subscribe(topics);
      client.on('message', (topic: string, message: Buffer, packet: any) => {
                const mess = message.toString();
                this.log("MQTT state message received:", mess);
                const name = mess.replace('_',' ');
                this.name = name
                this.log("Motion Camera:", this.name);
                this.mqttHandler();
            })`

File: src/streamingDelegate.ts [hunk 1]
```diff
@@ -66,8 +78,24 @@ export class StreamingDelegate implements CameraStreamingDelegate {
...
+    }
     this.debug = this.ffmpegOpt.debug;
-
+    this.additionalCommandline = this.ffmpegOpt.additionalCommandline || '-tune zerolatency';
+    this.vflip = this.ffmpegOpt.vflip || false;
+    this.hflip = this.ffmpegOpt.hflip || false;
+    this.mapvideo = this.ffmpegOpt.mapvideo || "0:0";
+    this.mapaudio = this.ffmpegOpt.mapaudio || "0:1";
+    this.videoFilter = this.ffmpegOpt.videoFilter || null; // null is a valid discrete value
+    this.debug = this.ffmpegOpt.debug;
```
Comment (longzheng) [inline src/streamingDelegate.ts:97]: This seems to be a duplicate. There's already a `this.debug` declaration on line 90.

File: src/streamingDelegate.ts [hunk 2]
```diff
@@ -66,8 +78,24 @@ export class StreamingDelegate implements CameraStreamingDelegate {
...
+      this.minBitrate = this.maxBitrate;
+    }
     this.debug = this.ffmpegOpt.debug;
-
+    this.additionalCommandline = this.ffmpegOpt.additionalCommandline || '-tune zerolatency';
+    this.vflip = this.ffmpegOpt.vflip || false;
+    this.hflip = this.ffmpegOpt.hflip || false;
+    this.mapvideo = this.ffmpegOpt.mapvideo || "0:0";
+    this.mapaudio = this.ffmpegOpt.mapaudio || "0:1";
+    this.videoFilter = this.ffmpegOpt.videoFilter || null; // null is a valid discrete value
```
Comment (longzheng) [inline src/streamingDelegate.ts:96]: Most of these properties don't seem to be used in the actual methods below.

For example, `this.fps` is not used in `handleStreamRequest`. It fetches the `this.ffmpegOpt.maxFPS` directly
https://github.com/homebridge-plugins/homebridge-camera-ffmpeg/blob/5ae442f78ec85b5647fee0f3cc9eecabf22172d1/src/streamingDelegate.ts#L252

**Comment by fennec622** [comment]
yes you can exclure 

, packet: any

File: src/index.ts [hunk 3]
```diff
@@ -184,19 +184,17 @@ class FfmpegPlatform implements DynamicPlatformPlugin {
...
         this.log('MQTT CONNECTED!');
       });
       client.subscribe(topics);
-      client.on('message', (topic: string) => {
-        const status = topic.toString();
-        const parts = status.split('/');
-        const partsThree = parts[2];
-        this.log('MQTT state message received:', status);
-        const name = partsThree.replace('_', ' ');
+      client.on('message', (topic: string, message: Buffer, packet: any) => {
```
[the remark this answers is not in the document]
Comment (fennec622) [reply src/index.ts:205]: yes you can exclure

, packet: any

**Comment by fennec622** [comment]
> @fennec622 last thing. I have added a few things into this for a stateless doorbell button. Also have an option to disable the fake switches so they are not there. those shouldn't be needed when using mqtt right?

You can disable Fake switch

You must change 

'const motionSwitch = accessory.getService(hap.Service.Switch);
        if (motionSwitch) {
          motionSwitch.setCharacteristic(hap.Characteristic.On, true);
        }
'
To
'const motionSenSor = accessory.getService(hap.Service.MotionSensor);
        if (motionSenSor) {
          motionSenSor.setCharacteristic(hap.Characteristic.MotionDetected, 1);
        }
'

But don't forget to motionSwitch.setCharacteristic(hap.Characteristic.MotionDetected, 0) after timeout

**Comment by donavanbecker** [comment]
> But don't forget to motionSwitch.setCharacteristic(hap.Characteristic.MotionDetected, 0) after timeout
What do you mean by that?

**Comment by fennec622** [comment]
> > But don't forget to motionSwitch.setCharacteristic(hap.Characteristic.MotionDetected, 0) after timeout
> 
> What do you mean by that?
> 
> I have updated the other stuff!

I see new index.ts

Now you have 4 types of button

},
          "motion": {
            "title": "Enable Motion",
            "type": "boolean"
          },
          "doorbell": {
            "title": "Enable Doorbell",
            "type": "boolean"
          },
          "doorbellSwitch": {
            "title": "Enable Doorbell Button",
            "type": "boolean"
          },
          "switches": {
            "title": "Enable Manual Automation Motion and/or Doorbell Switch",
            "type": "boolean"
          },

In mqtt you must put 

mqttHandler(): void {
    this.accessories.forEach((accessory: PlatformAccessory) => {
      if (accessory.displayName == this.name) {
        this.log('Switch Motion Detect On :', accessory.displayName);
        const motionSenSor = accessory.getService(hap.Service.MotionSensor);
        const doorbellSenSor = accessory.getService(hap.Service.Doorbell);
        if (motionSenSor) {
          motionSenSor.setCharacteristic(hap.Characteristic.MotionDetected, 1);
        }
        if (doorbellSenSor) {
          doorbellSenSor.updateCharacteristic(
                  hap.Characteristic.ProgrammableSwitchEvent,
                  hap.Characteristic.ProgrammableSwitchEvent.SINGLE_PRESS,
                );
        }
      }
    });
  }

but when you have just motion with no switch, motion always on because no timeout, you can put timeout in mqttHandler

if (cameraConfig.motion) {
      const motion = new hap.Service.MotionSensor(cameraConfig.name);
      cameraAccessory.addService(motion);
      const log = this.log;

      if (cameraConfig.switches) {
        const button = new hap.Service.Switch(cameraConfig.name, 'MotionTrigger');
        button
          .getCharacteristic(hap.Characteristic.On)
          .on(CharacteristicEventTypes.SET, (on: CharacteristicValue, callback: CharacteristicSetCallback) => {
            log.info(`Setting ${cameraAccessory.displayName} Motion to ${on}`);
            const motionService = cameraAccessory.getService(hap.Service.MotionSensor);
            if (motionService) {
              motionService.setCharacteristic(hap.Characteristic.MotionDetected, on ? 1 : 0);
              if (on) {
                setTimeout(function () {
                  log.info(`Setting ${cameraAccessory.displayName} Button to false`);
                  const switchService = cameraAccessory.getService(hap.Service.Switch);
                  if (switchService) {
                    switchService.setCharacteristic(hap.Characteristic.On, false);
                  }
                }, 5000);
              }
            }
            callback(null, on);
          });
        cameraAccessory.addService(button);
      }
    }

**Comment by donavanbecker** [comment]
> I see new index.ts

I am not following you, can you just submit a PR on to the beta.

**Comment by fennec622** [comment]
I try if all is ok and i submit

**Comment by donavanbecker** [comment]
> I try if all is ok and i submit

We have made some changes so that mqtt doesn't run unless it has config.
Just so you are aware.

**Comment by fennec622** [comment]
I make change for doorbell and motion 
Tested all is ok

**Comment by donavanbecker** [comment]
@california444, @Sunoo, @fennec622, and @longzheng, Everything seems to be running really good on my end. 

Any thing else you guys would like to see before we push v2.1.0 out?

**Comment by Sunoo** [comment]
@donavanbecker I'm good for now.

At some point I'd like to fix (unless someone else beats me to it) the configuration for a camera not updating once it's cached, but that is an existing problem, so isn't urgent.

**Comment by california444** [comment]
Let‘s go from my side...

**Pull request merged**